### PR TITLE
Make SelectedDeclarationProvider argument-aware

### PR DESCRIPTION
**Pull request by retailcoder**
This pull requests makes the `SelectionService` resolve the selected `ParameterDeclaration` given a selection that's inside an `ArgumentContext`; the parameter declaration is used for the "current selection" caption when the argument is an expression that doesn't otherwise resolve to another declaration, like when an argument is a literal:

literal string argument to MsgBox function shows it's the Prompt parameter and it's a Variant

Or when the argument is an expression with late-bound elements selected:

Sheet1.Cells arguments are late-bound per the implicit default member call

But any early-bound more specific sub-expression that resolves to a declaration remains the caption for other parameters:

image

Named arguments in a different order than defined, doesn't break:

image

Missing arguments aren't handled and behave as previously:

image

### CODE REVIEW

**Comment by retailcoder** [comment]
Tweaking "find all references" to locate argument refs for parameter declarations would be a nice next step.

**Comment by bclothier** [comment]
Sweet enhancement!

2 small comments based on the images:

1) You probably may want to have it display `Object`, rather than `IDispatch` (2nd image) as that is more familiar to the users. 

2) I am not sure I agree that missing argument should be skipped. I don't know if it's easy/feasible but I think I would rather see the missing argument name & type shown as a hint. Think about how many time people made a mistake of `MsgBox "foo", "my title"`....

File: Rubberduck.Parsing/VBA/SelectedDeclarationProvider.cs [hunk 1]
```diff
@@ -93,6 +99,21 @@ public Declaration SelectedDeclaration(QualifiedSelection qualifiedSelection)
             return SelectedModule(qualifiedSelection);
         }
 
+        private static ParameterDeclaration SelectedDeclarationViaArgument(QualifiedSelection qualifiedSelection, DeclarationFinder finder)
+        {
+            var members = finder.Members(qualifiedSelection.QualifiedName).Where(m => m.DeclarationType.HasFlag(DeclarationType.Procedure));
```
Comment (BZngr) [inline Rubberduck.Parsing/VBA/SelectedDeclarationProvider.cs:104]: It looks like selected arguments from Function and Properties are being excluded.  If that is not the intent, wouldn't
 `DeclarationType.HasFlag(DeclarationType.Member)` be correct here?

Comment (retailcoder) [reply Rubberduck.Parsing/VBA/SelectedDeclarationProvider.cs:104]: At that stage in the procedure the idea is to identify the scope we're in, so excluding parameters here, is intentional.

**Comment by retailcoder** [comment]
@bclothier 1 is a good idea, I'll do that. Hadn't realized 2 is possible when there are no named arguments, this would work too.

File: Rubberduck.Parsing/VBA/SelectedDeclarationProvider.cs (hunk 1 above)
Comment (BZngr) [reply Rubberduck.Parsing/VBA/SelectedDeclarationProvider.cs:104]: Perhaps I'm misunderstanding...Yes, parameters are excluded - but using `DeclarationType.Procedure` only includes Subroutines and Property Let/Set scopes. No?

**Comment by retailcoder** [comment]
Missing arguments work!

missing argument info

The argument context excludes the preceding whitespace, so I made a little dance and I'm treating the preceding whitespace token as part of the argument - this has the side-effect of lining up the "selected/current argument" with the VBE's parameter quick-info:

resolving to Buttons parameter despite being in the middle of whitespace and line continuations

quick-info showing the Buttons parameter in bold

**Comment by retailcoder** [comment]
IDispatch is now shown as 'Object' in the commandbar

**Comment by retailcoder** [comment]
Okay now this is getting pretty nice..

image

**Review by bclothier — APPROVED** [review]
LGTM, nice work! Only one little nitpick about magical literals.

File: Rubberduck.Core/UI/Command/MenuItems/CommandBars/IContextFormatter.cs [hunk 2]
```diff
@@ -125,9 +125,13 @@ private static string TypeName(Declaration declaration, bool multipleControls, s
                 return RubberduckUI.ContextMultipleControlsSelection;
             }
 
+            var friendlyTypeName = "IDispatch".Equals(declaration.AsTypeName, System.StringComparison.InvariantCultureIgnoreCase)
+                ? "Object"
```
[the remark this answers is not in the document]
Comment (retailcoder) [reply Rubberduck.Core/UI/Command/MenuItems/CommandBars/IContextFormatter.cs:129]: Absolutely, will do - I'm always a bit reluctant to add `using Rubberduck.Parsing.*` anywhere it isn't already though.

**Pull request merged**